### PR TITLE
Improve BQ <-> Avro conversions

### LINKED ISSUE #20677: Avro/BigQuery disagreement on DATE type representation
I'm trying to write a job using an Avro format function to write to BQ. The BQ table has a `DATE` column, Beam converts that to a `[ "null", "string" ]`, but when the BQ load job runs, it errors out with:

`Field date_column has incompatible types. Configured schema: date; Avro file: string.`

If I write an integer, I get a `Not in union ["null","string"]` error from Avro.

This happens regardless of whether `useAvroLogicalTypes` is set.

Imported from Jira [BEAM-11001](https://issues.apache.org/jira/browse/BEAM-11001). Original Jira may contain additional context.
Reported by: Hale.

### PR DESCRIPTION
**Pull request by RustedBones**
Extracted from #32360

This change-set focuses on conversions between BQ `TableSchema`/`TableRow` and Avro `Schema`/`GenericRecord`

Fix #20677

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java [hunk 1]
```diff
@@ -446,10 +456,27 @@ public static Schema fromTableSchema(TableSchema tableSchema, SchemaConversionOp
...
   public static org.apache.avro.Schema toGenericAvroSchema(
       String schemaName, List<TableFieldSchema> fieldSchemas) {
-    return BigQueryAvroUtils.toGenericAvroSchema(schemaName, fieldSchemas);
+    return toGenericAvroSchema(schemaName, fieldSchemas, false);
+  }
+
+  /** Convert a list of BigQuery {@link TableFieldSchema} to Avro {@link org.apache.avro.Schema}. */
+  public static org.apache.avro.Schema toGenericAvroSchema(
+      String schemaName, List<TableFieldSchema> fieldSchemas, Boolean stringLogicalTypes) {
+    return BigQueryAvroUtils.toGenericAvroSchema(schemaName, fieldSchemas, stringLogicalTypes);
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java:479]: Consider sync the parameter name `stringLogicalTypes` -> `useAvroLogicalTypes` ?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 2]
```diff
@@ -1275,8 +1272,12 @@ public PCollection<T> expand(PBegin input) {
 
       Schema beamSchema = null;
       if (getTypeDescriptor() != null && getToBeamRowFn() != null && getFromBeamRowFn() != null) {
-        beamSchema = sourceDef.getBeamSchema(bqOptions);
-        beamSchema = getFinalSchema(beamSchema, getSelectedFields());
+        TableSchema tableSchema = sourceDef.getTableSchema(bqOptions);
+        ValueProvider<List<String>> selectedFields = getSelectedFields();
+        if (selectedFields != null) {
+          tableSchema = BigQueryUtils.trimSchema(tableSchema, selectedFields.get());
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:1278]: We still need to check `selectedFields.isAccessible()` (as the original getFinalSchema) as this is done in pipeline expansion time

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceDef.java [hunk 3]
```diff
@@ -45,6 +45,15 @@ <T> BigQuerySourceBase<T> toSource(
...
 
+  /**
+   * Extract the {@link TableSchema} corresponding to this source.
+   *
+   * @param bqOptions BigQueryOptions
+   * @return table schema of the source
+   * @throws BigQuerySchemaRetrievalException if schema retrieval fails
+   */
+  TableSchema getTableSchema(BigQueryOptions bqOptions);
+
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceDef.java:56]: Now getBeamSchema is trivial and the impl in its implementation classes are the same. Consider make a default implementation here?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java [hunk 4]
```diff
@@ -64,35 +58,95 @@
...
+        // decimal
+        LogicalType logicalType;
+        if (schema.getScale() != null) {
+          logicalType =
+              LogicalTypes.decimal(schema.getPrecision().intValue(), schema.getScale().intValue());
+        } else if (schema.getPrecision() != null) {
+          logicalType = LogicalTypes.decimal(schema.getPrecision().intValue());
+        } else if (bqType.equals("NUMERIC")) {
+          logicalType = LogicalTypes.decimal(38, 9);
+        } else {
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java:114]: add a comment `// BIGNUMERIC`

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java [hunk 5]
```diff
@@ -193,8 +193,7 @@ public void teardown() {
   @Test
   public void testBuildTableBasedSource() {
     BigQueryIO.TypedRead<TableRow> typedRead =
-        BigQueryIO.read(new TableRowParser())
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOStorageReadTest.java:196]: same, test still pass on old config. We can preserve the old test code (or just update some of them) so it still covers different config settings.

**Review by Abacn — COMMENTED** [review]
Thanks for this effort, went through once and had comments. For BQIO changes, @ahmedabu98 could you please also take a look? Thanks!

**Review by ahmedabu98 — COMMENTED** [review]
+1 thanks for this effort. Left a few comments but the rest looks great

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java [hunk 6]
```diff
@@ -273,145 +292,104 @@ private static List<Object> convertRepeatedField(
...
+          // SQL types TIME
           return formatTime((Long) v);
+        } else if (logicalType instanceof LogicalTypes.TimestampMicros) {
+          // SQL types TIMESTAMP
+          return formatTimestamp((Long) v);
         } else {
-          verify(v instanceof CharSequence, "Expected CharSequence (String), got %s", v.getClass());
-          return v.toString();
+          // SQL types INT64 (INT, SMALLINT, INTEGER, BIGINT, TINYINT, BYTEINT)
+          return ((Long) v).toString();
```
[the remark this answers is not in the document]
Comment (RustedBones) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java:369]: Yes, tests are failing.
I tried to check what the java BQ client returns when querying a table, and it looks consistent (`TableRow` record from `bigquery-public-data.samples.shakespeare`)
image

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java [hunk 7]
```diff
@@ -64,35 +60,96 @@
...
+        // double
+        return SchemaBuilder.builder().doubleType();
+      case "BOOLEAN":
+      case "BOOL":
+        // boolean
+        return SchemaBuilder.builder().booleanType();
+      case "TIMESTAMP":
+        // in Extract Jobs, it always uses the Avro logical type
+        // we may have to change this if we move to EXPORT DATA
+        return LogicalTypes.timestampMicros().addToSchema(SchemaBuilder.builder().longType());
```
Comment (RustedBones) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java:106]: `useAvroLogicalTypes` has a different behavior when used in extract job vs export data for `TIMESTAMP` columns.

Currently, beam's implementation of query is relying on temp table + extract job, so we don't have to handle this discrepancy.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java:364]: Sounds good, thanks for checking

**Review by ahmedabu98 — APPROVED** [review]
LGTM

**Comment by ahmedabu98** [comment]
Failing test is unrelated, merging now

**Pull request merged**